src/runtime/hexagon_remote/log.cpp

```diff
@@ -0,0 +1,75 @@
+#include "log.h"
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <qurt.h>
+
+// On device, we make a circular buffer that log_printf writes to, and
+// we can read from in a remote RPC call.
+class Log {
+    char *buffer;
+    int size;
+    int read_cursor;
+    int write_cursor;
+
+    qurt_mutex_t lock;
+
+public:
+    // The size must be a power of 2.
+    Log(int size) : buffer(NULL), size(size), read_cursor(0), write_cursor(0) {
+        qurt_mutex_init(&lock);
```

Comment:
The code below uses the and operator to implement a circular buffer.

Even if we have an assert... if it fails, then it will be impossible to get a message out of the system :) If this class doesn't successfully construct, then there's no way to print anything.